doc/TACACSPLUS_PASSKEY_ENCRYPTION.md

```diff
@@ -0,0 +1,111 @@
+# TACACS+ Passkey Encryption #
+
+
+## Table of Contents
+
+- [Revision](#revision)
+- [Scope](#scope)
+- [Abbreviations](#abbreviations)
+- [Overview](#overview)
+- [Requirements](#requirements)
+- [High-Level Design](#high-level-design)
+- [Implementation Details](#implementation-details)
+        - [Show CLI changes](@show-cli-changes)
+- [Benifits](#benifits)
+- [Testing Requirements](#testing-requirements)
+
+
+
+### Revision
+
+ | Rev |     Date    |       Author         | Change Description                |
+ |:---:|:-----------:|:--------------------:|-----------------------------------|
+ | 0.1 |             | Nikhil Moray (nmoray)| Initial version                   |
+ | 0.1 |             | Madhu Paluru (madhupalu)| Updated                        |
+
+
+ ### Scope
+
+This document describes the High Level Design of "TACACS+ Passkey Encryption" feature support in SONiC. It encompasses design and implementation considerations for enhancing the security of TACACS+ (Terminal Access Controller Access-Control System) authentication by introducing an encrypted passkey.
+
+
+### Abbreviations
+
+ | Term    |     Meaning                                                        |
+ |:-------:|:-------------------------------------------------------|
+ | TACACS  | Terminal Access Controller Access Control System Plus) |
+
+### Overview
+
+This addition constitutes a substantial improvement in bolstering the security of the TACACS+ authentication protocol. TACACS+ has a well-established reputation as a reliable means of managing access to network devices. However, the previous practice of utilising a sensitive passkey in plaintext during the authentication process posed security concerns. With this enhancement, the vulnerability is effectively mitigated by introducing robust passkey encryption mechanisms (on the client side), ensuring the safeguarding of authentication credentials and an overall strengthening of network security."
+
+
+### Requirements
+
+The primary objective of this feature is to safeguard the TACACS passkey, which is stored in its plaintext format within CONFIG_DB.
```

Comment:
Can you add the section for DB migrator?

Reply:
Can you add the YANG change for the new TACACS passkey?

Reply:
@venkatmahalingam passkey is already part of yang model, no new schema introduced part of HLD - https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-system-tacacs.yang
Only the backend will use system mac as unique salt to encrypt/decrypt the passkey.

Reply:
We don't need DB migrator as no change to the schema.

Reply:
How about existing plaintext passkey to encrypted key migration in the config-db during image upgrade?

Reply:
We should update the YANG to reflect the plaintext to encrypt key change in the passkey description.

Reply:
@venkatmahalingam ok, will take care of yang model change and added db migration section in the HLD.

Reply:
@venkatmahalingam The current yang model **description already reflect the change you requested** - '   description "Shared secret used for encrypting the communication";'  Is the sufficient or do we think still need an update? 

 leaf passkey {
                    type string {
                        length "1..65";
                        pattern "[^ #,]*" {
                            error-message 'TACACS shared secret (Valid chars are ASCII printable except SPACE, "#", and ",")';
                        }
                    }
                    description "Shared secret used for encrypting the communication";
                }

Reply:
@venkatmahalingam is [sonic-system-tacacs.yang](https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-system-tacacs.yang) a pull model or push? If we need to configure the passkey via this model, we don't need any change as it will still be in a plaintext format only.

Reply:
@nmoray , I think the length part need change, because when you encrypt a text to binary then convert to text, the length will usually be much longer, currently sonic already enable yang validation, if the length not change, yang validation will failed with some TACACS key.

Reply:
@liuh-80 one question related to sonic-system-tacacs.yang model. Is it being used for pushing the TACACS configs or pulling the telemetry data related to TACACS or both?

Reply:
sonic-system-tacacs.yang is used for push TACACS config, also when change TACACS config on device with CLI, the model also will be used for validation.

Increase the length of passkey is backward compatibility. No need other change.